blog-content/aws-firehose-opentelemetry-collector.md

```diff
@@ -0,0 +1,183 @@
+
+To use the contrib image, modify your OpenTelemetry Collector configuration as follows:
+
+```yaml
+receivers:
+    awsfirehose/cwmetrics:
+        endpoint: '0.0.0.0:4433'
+        record_type: cwmetrics
+        include_metadata: true
+    awsfirehose/cwlogs:
+        endpoint: '0.0.0.0:4434'
+        record_type: cwlogs
+        include_metadata: true
+    awsfirehose/otlp_v1:
+        endpoint: '0.0.0.0:4435'
+        record_type: otlp_v1
+        include_metadata: true
+    
+exporters:
+    logging:
+    verbosity: detailed
+    
+service:
+    pipelines:
+        logs:
+        receivers: [awsfirehose/cwmetrics, awsfirehose/cwlogs, awsfirehose/otlp_v1]
+        exporters: [logging]
+```
+
+This configuration sets up a Firehose receiver that listens on port 8000 and processes logs in JSON format.
+
+OpenTelemetry Firehose Receiver
+
+If a native OpenTelemetry Firehose receiver is unavailable or does not fully meet your needs, you may need to manually implement a receiver.
+
+Manual Implementation
+
+Custom parsing of Firehose logs can be done using a processor within OpenTelemetry. For example, using a custom script to preprocess logs before ingestion:
+
+```golang
+package http
+
+import (
+	"compress/gzip"
+	"encoding/base64"
+	"encoding/json"
+	"io"
+	"net/http"
+	"strings"
+)
+
+type FirehosePayload struct {
+	RequestId string
+	Timestamp int64
+	Records   []struct {
+		Data string
+	}
+}
+
+func ExtractFirehoseMetadata(r *http.Request, body []byte) (*FirehosePayload, [][]byte, error) {
+	var payload FirehosePayload
+	if err := json.Unmarshal(body, &payload); err != nil {
+		return nil, nil, err
+	}
+
+	attributesMap := struct {
+		CommonAttributes struct {
+			CustomHeader string `json:"x-custom-header"`
+		} `json:"commonAttributes"`
+	}{}
+	if err := json.Unmarshal([]byte(r.Header.Get("X-Amz-Firehose-Common-Attributes")), &attributesMap); err != nil {
+		return nil, nil, err
+	}
+
+	rawRecords := make([][]byte, 0, len(payload.Records))
+	for _, l := range payload.Records {
+		data, err := base64.StdEncoding.DecodeString(l.Data)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		var msg []byte
+		// try to load data as gzip. if it is not, assume it is not compressed
+		gz, err := gzip.NewReader(strings.NewReader(string(data)))
+		if err == nil {
+			msg, err = io.ReadAll(gz)
+			if err != nil {
+				return nil, nil, err
+			}
+		} else {
+			msg = data
+		}
+		rawRecords = append(rawRecords, msg)
+	}
+
+	return &payload, rawRecords, nil
+}
+```
+
+This function extracts key log attributes and reformats them for further processing in OpenTelemetry.
+
+## Conclusion
+
+Integrating AWS Firehose with OpenTelemetry provides a powerful way to collect and analyze telemetry data from AWS services. Whether you use a native receiver or implement a custom solution, Firehose enables seamless data streaming for improved observability.
```

Comment:
WDYT about adding something about how leveraging existing receivers can streamline the process of collecting telemetry data and hit on some of the benefits of using a receiver vs other approaches?

Reply:
added to the otel paragraph! lmk what you think

Reply:
Nice, I like it! One minor thing is that I feel like the last paragraph is a little duplicative - maybe unnecessary?>